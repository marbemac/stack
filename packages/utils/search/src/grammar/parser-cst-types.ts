/* eslint-disable @typescript-eslint/consistent-type-definitions */

// generated via yarn build.grammar-types

import type { CstNode, ICstVisitor, IToken } from "chevrotain";

export interface SearchQueryCstNode extends CstNode {
  name: "searchQuery";
  children: SearchQueryCstChildren;
}

export type SearchQueryCstChildren = {
  selectClause: SelectClauseCstNode[];
  fromClause: FromClauseCstNode[];
  whereClause?: WhereClauseCstNode[];
};

export interface SelectClauseCstNode extends CstNode {
  name: "selectClause";
  children: SelectClauseCstChildren;
}

export type SelectClauseCstChildren = {
  Select: IToken[];
  Identifier: IToken[];
  Comma?: IToken[];
};

export interface FromClauseCstNode extends CstNode {
  name: "fromClause";
  children: FromClauseCstChildren;
}

export type FromClauseCstChildren = {
  From: IToken[];
  Identifier: IToken[];
};

export interface WhereClauseCstNode extends CstNode {
  name: "whereClause";
  children: WhereClauseCstChildren;
}

export type WhereClauseCstChildren = {
  Where: IToken[];
  whereExpression: WhereExpressionCstNode[];
};

export interface WhereExpressionCstNode extends CstNode {
  name: "whereExpression";
  children: WhereExpressionCstChildren;
}

export type WhereExpressionCstChildren = {
  conditions?: (FilterCstNode | AtomicFilterValCstNode)[];
};

export interface FilterCstNode extends CstNode {
  name: "filter";
  children: FilterCstChildren;
}

export type FilterCstChildren = {
  Negate?: IToken[];
  lhs?: (FunctionCstNode | FilterKeyCstNode)[];
  filterOp: FilterOpCstNode[];
  rhs: FilterValCstNode[];
};

export interface FunctionCstNode extends CstNode {
  name: "function";
  children: FunctionCstChildren;
}

export type FunctionCstChildren = {
  Identifier: IToken[];
  LParen: IToken[];
  functionArg?: FunctionArgCstNode[];
  Comma?: IToken[];
  RParen: IToken[];
};

export interface FunctionArgCstNode extends CstNode {
  name: "functionArg";
  children: FunctionArgCstChildren;
}

export type FunctionArgCstChildren = {
  args?: (FilterCstNode | AtomicFilterValCstNode)[];
  WhiteSpace?: IToken[];
};

export interface FilterKeyCstNode extends CstNode {
  name: "filterKey";
  children: FilterKeyCstChildren;
}

export type FilterKeyCstChildren = {
  Identifier: IToken[];
};

export interface FilterValCstNode extends CstNode {
  name: "filterVal";
  children: FilterValCstChildren;
}

export type FilterValCstChildren = {
  val?: (FilterInCstNode | RelativeDateFilterValCstNode | AtomicFilterValCstNode)[];
};

export interface FilterInCstNode extends CstNode {
  name: "filterIn";
  children: FilterInCstChildren;
}

export type FilterInCstChildren = {
  LBracket: IToken[];
  atomicFilterVal: AtomicFilterValCstNode[];
  Comma?: IToken[];
  RBracket: IToken[];
};

export interface RelativeDateFilterValCstNode extends CstNode {
  name: "relativeDateFilterVal";
  children: RelativeDateFilterValCstChildren;
}

export type RelativeDateFilterValCstChildren = {
  op?: (IToken)[];
  Number: IToken[];
  DateUnit: IToken[];
};

export interface AtomicFilterValCstNode extends CstNode {
  name: "atomicFilterVal";
  children: AtomicFilterValCstChildren;
}

export type AtomicFilterValCstChildren = {
  LQuote?: IToken[];
  QuotedIdentifier?: IToken[];
  RQuote?: IToken[];
  Identifier?: IToken[];
  Number?: IToken[];
};

export interface FilterOpCstNode extends CstNode {
  name: "filterOp";
  children: FilterOpCstChildren;
}

export type FilterOpCstChildren = {
  Colon: IToken[];
  op?: (IToken)[];
};

export interface TSearchCstVisitor<IN, OUT> extends ICstVisitor<IN, OUT> {
  searchQuery(children: SearchQueryCstChildren, param?: IN): OUT;
  selectClause(children: SelectClauseCstChildren, param?: IN): OUT;
  fromClause(children: FromClauseCstChildren, param?: IN): OUT;
  whereClause(children: WhereClauseCstChildren, param?: IN): OUT;
  whereExpression(children: WhereExpressionCstChildren, param?: IN): OUT;
  filter(children: FilterCstChildren, param?: IN): OUT;
  function(children: FunctionCstChildren, param?: IN): OUT;
  functionArg(children: FunctionArgCstChildren, param?: IN): OUT;
  filterKey(children: FilterKeyCstChildren, param?: IN): OUT;
  filterVal(children: FilterValCstChildren, param?: IN): OUT;
  filterIn(children: FilterInCstChildren, param?: IN): OUT;
  relativeDateFilterVal(children: RelativeDateFilterValCstChildren, param?: IN): OUT;
  atomicFilterVal(children: AtomicFilterValCstChildren, param?: IN): OUT;
  filterOp(children: FilterOpCstChildren, param?: IN): OUT;
}
