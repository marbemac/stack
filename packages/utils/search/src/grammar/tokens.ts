import { createToken, Lexer } from 'chevrotain';

/**
 * @example
 * plan:
 * foo.bar:
 * $last_seen:
 * company.$id:
 */
export const QualifierKey = createToken({ name: 'QualifierKey', pattern: /[@_A-Za-z$][a-zA-Z0-9@_$\-.]*:/ });

export const Identifier = createToken({ name: 'Identifier', pattern: /[@_A-Za-z0-9$][a-zA-Z0-9@_$\-.]*/ });

// Quoted identifiers support anything except for quotes
export const QuotedIdentifier = createToken({ name: 'QuotedIdentifier', pattern: /[^"]+/ });

/**
 * The longer_alt here allows us to handle strings that start with a number, e.g. "3industries"
 */
export const Number = createToken({ name: 'Number', pattern: /-?(0|[1-9]\d*)(\.\d+)?/, longer_alt: Identifier });

export const Iso8601Date = createToken({
  name: 'Iso8601Date',
  pattern: /\d{4}(-\d\d(-\d\d(T\d\d:\d\d(:\d\d)?(\.\d+)?(([+-]\d\d:\d\d)|Z)?)?)?)?/i,
  // The longer alts here are in case of large numbers that might trigger the Iso check, and fallback to plain text if nothing matches
  longer_alt: [Number, Identifier],
});

// We specify the "longer_alt" property to resolve keywords vs identifiers ambiguity.
// See: https://github.com/chevrotain/chevrotain/blob/master/examples/lexer/keywords_vs_identifiers/keywords_vs_identifiers.js
export const Select = createToken({
  name: 'Select',
  pattern: /SELECT/,
  longer_alt: Identifier,
});

export const From = createToken({
  name: 'From',
  pattern: /FROM/,
  longer_alt: Identifier,
});

export const Where = createToken({
  name: 'Where',
  pattern: /WHERE/,
  longer_alt: Identifier,
});

export const LParen = createToken({ name: 'LParen', pattern: /\(/ });
export const RParen = createToken({ name: 'RParen', pattern: /\)/ });
export const LBracket = createToken({ name: 'LBracket', pattern: /\[/ });
export const RBracket = createToken({ name: 'RBracket', pattern: /]/ });
export const Comma = createToken({ name: 'Comma', pattern: /,\s*/ });
export const Colon = createToken({ name: 'Colon', pattern: /:/ });
export const Greater = createToken({ name: 'Greater', pattern: />/ });
export const Less = createToken({ name: 'Less', pattern: /</ });
export const GreaterEq = createToken({ name: 'GreaterEq', pattern: />=/ });
export const LessEq = createToken({ name: 'LessEq', pattern: /<=/ });
export const Boolean = createToken({ name: 'Boolean', pattern: /true|false/ });
export const Plus = createToken({ name: '+', pattern: /\+/ });
export const Minus = createToken({ name: 'Minus', pattern: /-/ });
export const Equals = createToken({ name: 'Equals', pattern: /=/ });
export const Negate = createToken({ name: 'Negate', pattern: /!/ });

// https://chevrotain.io/docs/guide/custom_token_patterns.html#custom-payloads
// We're using this to get access to the regex capture groups in payload later (in the visitor)
export interface RelativeDateTokenPayload {
  sign: '+' | '-';
  value: string;
  unit: 's' | 'mi' | 'h' | 'd' | 'w' | 'm' | 'y';
}
const relativeDatePattern = /(?<sign>-|\+)(?<value>\d+)(?<unit>[s|mi|h|d|w|m|q|y])/y;
export const RelativeDate = createToken({
  name: 'RelativeDate',
  pattern: (text, startOffset) => {
    relativeDatePattern.lastIndex = startOffset;

    const result = relativeDatePattern.exec(text);
    if (result) {
      // @ts-expect-error ignore - this is the official way to add custom payload to tokens w chevrotain
      result.payload = result.groups;
    }

    return result;
  },
});

export const WhiteSpace = createToken({
  name: 'WhiteSpace',
  pattern: /\s+/,
  group: Lexer.SKIPPED,
});

export const LQuote = createToken({ name: 'LQuote', pattern: /"/, push_mode: 'quoted_mode' });
export const RQuote = createToken({ name: 'RQuote', pattern: /"/, pop_mode: true });
