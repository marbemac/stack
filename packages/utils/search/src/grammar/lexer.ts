import { type ILexerConfig, type IMultiModeLexerDefinition, Lexer } from 'chevrotain';

import * as t from './tokens.ts';

// note we are placing WhiteSpace first as it is very common thus it will speed up the lexer.
export const lexerTokenDef: IMultiModeLexerDefinition = {
  defaultMode: 'main',
  modes: {
    main: [
      t.WhiteSpace,

      // "keywords" appear before the Identifier
      t.Select,
      t.From,
      t.Where,
      t.LParen,
      t.RParen,
      t.LBracket,
      t.RBracket,
      t.LQuote,
      t.Comma,
      t.Colon,
      t.GreaterEq,
      t.Greater,
      t.LessEq,
      t.Less,
      t.Plus,
      t.Minus,
      t.Equals,
      t.Negate,

      // The Identifier must appear after the keywords because all keywords are valid identifiers.
      t.Identifier,

      t.Integer,
    ],

    quoted_mode: [t.RQuote, t.QuotedIdentifier],
  },
};

interface SearchLexerOpts extends ILexerConfig {}

export const createSearchLexer = (opts?: SearchLexerOpts) => {
  return new Lexer(lexerTokenDef, { positionTracking: 'onlyOffset', ...opts });
};
