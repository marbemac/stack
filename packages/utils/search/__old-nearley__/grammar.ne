@preprocessor typescript

@{%

import { createLexer } from './lexer.js';

const lexer = createLexer();

%}

# https://omrelli.ug/nearley-playground/ is helpful to experiment
# Required adjustments to the content of this file if copy pasting over to playground:
#   - remove "@preprocessor typescript" statement on line ~1
#   - paste contents of `lexer.ts` over the `import ... ./lexer.js` statement on line ~5
#   - remove all typescript typings from the copied code (easiest is to copy into JS file to highlight and delete)
#   - change `import moo from 'moo';` to `const moo = require('moo')` (es6 imports not supported)

@lexer lexer

Main -> (Filter | _ | Raw):*  {% transformMain %}

Filter ->
	TextFilter {% id %}
	| QuotedTextFilter {% id %}
	| TextFilterIn {% id %}

TextFilter -> %negation:? %qualKey %qualOp:? %qualVal:? {% ([negated,key,op,val]) => ({ negated: Boolean(negated), key, op: op || '=', val: val || '' }) %}

QuotedTextFilter -> %negation:? %qualKey %quote %qualVal:? %quote:? {% ([negated,key,,val]) => ({ negated: Boolean(negated), key, op: '=', val: val || '' }) %}

TextFilterIn -> %negation:? %qualKey %lbracket QualMultiVal:? %rbracket:? {% ([negated, key,,val]) => ({ negated: Boolean(negated), key, op: '=', val: val || '' }) %}

QualMultiVal -> (%qualVal %comma:?):+ {% ([val]) => val.map((v) => v[0]) %}

Raw -> %string {% id %}

_ -> %space {% () => null %}

@{%

function transformMain([tokens]) {
	const noFalsey = [];
	for (const f of tokens) {
		if (f[0]) {
			noFalsey.push(f[0])
		}
	}

  return {
		tokens: noFalsey
	};
}

%}
