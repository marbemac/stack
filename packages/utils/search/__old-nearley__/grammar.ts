/* eslint-disable */
// @ts-nocheck
// Generated automatically by nearley, version 2.20.1
// http://github.com/Hardmath123/nearley
// Bypasses TS6133. Allow declared but unused functions.
// @ts-ignore
function id(d: any[]): any { return d[0]; }
declare var negation: any;
declare var qualKey: any;
declare var qualOp: any;
declare var qualVal: any;
declare var quote: any;
declare var lbracket: any;
declare var rbracket: any;
declare var comma: any;
declare var string: any;
declare var space: any;


import { createLexer } from './lexer.js';

const lexer = createLexer();




function transformMain([tokens]) {
	const noFalsey = [];
	for (const f of tokens) {
		if (f[0]) {
			noFalsey.push(f[0])
		}
	}

  return {
		tokens: noFalsey
	};
}


interface NearleyToken {
  value: any;
  [key: string]: any;
};

interface NearleyLexer {
  reset: (chunk: string, info: any) => void;
  next: () => NearleyToken | undefined;
  save: () => any;
  formatError: (token: never) => string;
  has: (tokenType: string) => boolean;
};

interface NearleyRule {
  name: string;
  symbols: NearleySymbol[];
  postprocess?: (d: any[], loc?: number, reject?: {}) => any;
};

type NearleySymbol = string | { literal: any } | { test: (token: any) => boolean };

interface Grammar {
  Lexer: NearleyLexer | undefined;
  ParserRules: NearleyRule[];
  ParserStart: string;
};

const grammar: Grammar = {
  Lexer: lexer,
  ParserRules: [
    {"name": "Main$ebnf$1", "symbols": []},
    {"name": "Main$ebnf$1$subexpression$1", "symbols": ["Filter"]},
    {"name": "Main$ebnf$1$subexpression$1", "symbols": ["_"]},
    {"name": "Main$ebnf$1$subexpression$1", "symbols": ["Raw"]},
    {"name": "Main$ebnf$1", "symbols": ["Main$ebnf$1", "Main$ebnf$1$subexpression$1"], "postprocess": (d) => d[0].concat([d[1]])},
    {"name": "Main", "symbols": ["Main$ebnf$1"], "postprocess": transformMain},
    {"name": "Filter", "symbols": ["TextFilter"], "postprocess": id},
    {"name": "Filter", "symbols": ["QuotedTextFilter"], "postprocess": id},
    {"name": "Filter", "symbols": ["TextFilterIn"], "postprocess": id},
    {"name": "TextFilter$ebnf$1", "symbols": [(lexer.has("negation") ? {type: "negation"} : negation)], "postprocess": id},
    {"name": "TextFilter$ebnf$1", "symbols": [], "postprocess": () => null},
    {"name": "TextFilter$ebnf$2", "symbols": [(lexer.has("qualOp") ? {type: "qualOp"} : qualOp)], "postprocess": id},
    {"name": "TextFilter$ebnf$2", "symbols": [], "postprocess": () => null},
    {"name": "TextFilter$ebnf$3", "symbols": [(lexer.has("qualVal") ? {type: "qualVal"} : qualVal)], "postprocess": id},
    {"name": "TextFilter$ebnf$3", "symbols": [], "postprocess": () => null},
    {"name": "TextFilter", "symbols": ["TextFilter$ebnf$1", (lexer.has("qualKey") ? {type: "qualKey"} : qualKey), "TextFilter$ebnf$2", "TextFilter$ebnf$3"], "postprocess": ([negated,key,op,val]) => ({ negated: Boolean(negated), key, op: op || '=', val: val || '' })},
    {"name": "QuotedTextFilter$ebnf$1", "symbols": [(lexer.has("negation") ? {type: "negation"} : negation)], "postprocess": id},
    {"name": "QuotedTextFilter$ebnf$1", "symbols": [], "postprocess": () => null},
    {"name": "QuotedTextFilter$ebnf$2", "symbols": [(lexer.has("qualVal") ? {type: "qualVal"} : qualVal)], "postprocess": id},
    {"name": "QuotedTextFilter$ebnf$2", "symbols": [], "postprocess": () => null},
    {"name": "QuotedTextFilter$ebnf$3", "symbols": [(lexer.has("quote") ? {type: "quote"} : quote)], "postprocess": id},
    {"name": "QuotedTextFilter$ebnf$3", "symbols": [], "postprocess": () => null},
    {"name": "QuotedTextFilter", "symbols": ["QuotedTextFilter$ebnf$1", (lexer.has("qualKey") ? {type: "qualKey"} : qualKey), (lexer.has("quote") ? {type: "quote"} : quote), "QuotedTextFilter$ebnf$2", "QuotedTextFilter$ebnf$3"], "postprocess": ([negated,key,,val]) => ({ negated: Boolean(negated), key, op: '=', val: val || '' })},
    {"name": "TextFilterIn$ebnf$1", "symbols": [(lexer.has("negation") ? {type: "negation"} : negation)], "postprocess": id},
    {"name": "TextFilterIn$ebnf$1", "symbols": [], "postprocess": () => null},
    {"name": "TextFilterIn$ebnf$2", "symbols": ["QualMultiVal"], "postprocess": id},
    {"name": "TextFilterIn$ebnf$2", "symbols": [], "postprocess": () => null},
    {"name": "TextFilterIn$ebnf$3", "symbols": [(lexer.has("rbracket") ? {type: "rbracket"} : rbracket)], "postprocess": id},
    {"name": "TextFilterIn$ebnf$3", "symbols": [], "postprocess": () => null},
    {"name": "TextFilterIn", "symbols": ["TextFilterIn$ebnf$1", (lexer.has("qualKey") ? {type: "qualKey"} : qualKey), (lexer.has("lbracket") ? {type: "lbracket"} : lbracket), "TextFilterIn$ebnf$2", "TextFilterIn$ebnf$3"], "postprocess": ([negated, key,,val]) => ({ negated: Boolean(negated), key, op: '=', val: val || '' })},
    {"name": "QualMultiVal$ebnf$1$subexpression$1$ebnf$1", "symbols": [(lexer.has("comma") ? {type: "comma"} : comma)], "postprocess": id},
    {"name": "QualMultiVal$ebnf$1$subexpression$1$ebnf$1", "symbols": [], "postprocess": () => null},
    {"name": "QualMultiVal$ebnf$1$subexpression$1", "symbols": [(lexer.has("qualVal") ? {type: "qualVal"} : qualVal), "QualMultiVal$ebnf$1$subexpression$1$ebnf$1"]},
    {"name": "QualMultiVal$ebnf$1", "symbols": ["QualMultiVal$ebnf$1$subexpression$1"]},
    {"name": "QualMultiVal$ebnf$1$subexpression$2$ebnf$1", "symbols": [(lexer.has("comma") ? {type: "comma"} : comma)], "postprocess": id},
    {"name": "QualMultiVal$ebnf$1$subexpression$2$ebnf$1", "symbols": [], "postprocess": () => null},
    {"name": "QualMultiVal$ebnf$1$subexpression$2", "symbols": [(lexer.has("qualVal") ? {type: "qualVal"} : qualVal), "QualMultiVal$ebnf$1$subexpression$2$ebnf$1"]},
    {"name": "QualMultiVal$ebnf$1", "symbols": ["QualMultiVal$ebnf$1", "QualMultiVal$ebnf$1$subexpression$2"], "postprocess": (d) => d[0].concat([d[1]])},
    {"name": "QualMultiVal", "symbols": ["QualMultiVal$ebnf$1"], "postprocess": ([val]) => val.map((v) => v[0])},
    {"name": "Raw", "symbols": [(lexer.has("string") ? {type: "string"} : string)], "postprocess": id},
    {"name": "_", "symbols": [(lexer.has("space") ? {type: "space"} : space)], "postprocess": () => null}
  ],
  ParserStart: "Main",
};

export default grammar;
