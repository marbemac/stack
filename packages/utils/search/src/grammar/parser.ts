import {
  type CstNode,
  CstParser,
  type IParserConfig,
  type IRuleConfig,
  type ParserMethod,
  type TokenVocabulary,
} from 'chevrotain';

import { lexerTokenDef } from './lexer.ts';
import type {
  FromClauseCstNode,
  SearchQueryCstNode,
  SelectClauseCstNode,
  WhereClauseCstNode,
  WhereExpressionCstNode,
} from './parser-cst-types.ts';
import * as t from './tokens.ts';

class SearchParserImpl extends CstParser {
  constructor(tokenVocabulary: TokenVocabulary, config?: IParserConfig) {
    super(tokenVocabulary, config);
    this.performSelfAnalysis();
  }

  /** Pass through that just adjusts the typings to allow us to type the rule responses for the public api */
  #RULE<N extends CstNode, F extends () => void = () => void>(
    name: string,
    implementation: F,
    config?: IRuleConfig<N>,
  ): ParserMethod<Parameters<F>, N> {
    return this.RULE(name, implementation, config) as ParserMethod<Parameters<F>, N>;
  }

  public searchQuery = this.#RULE<SearchQueryCstNode>('searchQuery', () => {
    this.SUBRULE(this.selectClause);
    this.SUBRULE(this.fromClause);
    this.OPTION(() => {
      this.SUBRULE(this.whereClause);
    });
  });

  public selectClause = this.#RULE<SelectClauseCstNode>('selectClause', () => {
    this.CONSUME(t.Select);
    this.AT_LEAST_ONE_SEP({
      SEP: t.Comma,
      DEF: () => {
        this.CONSUME(t.Identifier);
      },
    });
  });

  public fromClause = this.#RULE<FromClauseCstNode>('fromClause', () => {
    this.CONSUME(t.From);
    this.CONSUME(t.Identifier);
  });

  public whereClause = this.#RULE<WhereClauseCstNode>('whereClause', () => {
    this.CONSUME(t.Where);
    this.SUBRULE(this.whereExpression);
  });

  public whereExpression = this.#RULE<WhereExpressionCstNode>('whereExpression', () => {
    this.MANY(() => {
      this.OR([
        { ALT: () => this.SUBRULE(this.#filter, { LABEL: 'conditions' }) },
        { ALT: () => this.SUBRULE(this.#atomicFilterVal, { LABEL: 'conditions' }) },
      ]);
    });
  });

  #filter = this.#RULE('filter', () => {
    this.OPTION(() => {
      this.CONSUME(t.Negate);
    });

    this.OR([
      { ALT: () => this.SUBRULE(this.#function, { LABEL: 'lhs' }) },
      { ALT: () => this.SUBRULE(this.#filterKey, { LABEL: 'lhs' }) },
    ]);

    this.SUBRULE(this.#filterOp);

    this.SUBRULE(this.#filterVal, { LABEL: 'rhs' });
  });

  #function = this.#RULE('function', () => {
    this.CONSUME(t.Identifier);
    this.CONSUME(t.LParen);

    this.MANY_SEP({
      SEP: t.Comma,
      DEF: () => {
        this.SUBRULE(this.#functionArg);
      },
    });

    this.CONSUME(t.RParen);
  });

  #functionArg = this.#RULE('functionArg', () => {
    this.MANY_SEP2({
      SEP: t.WhiteSpace,
      DEF: () => {
        this.OR([
          { ALT: () => this.SUBRULE(this.#filter, { LABEL: 'args' }) },
          { ALT: () => this.SUBRULE(this.#atomicFilterVal, { LABEL: 'args' }) },
        ]);
      },
    });
  });

  #filterKey = this.#RULE('filterKey', () => {
    this.CONSUME(t.Identifier);
  });

  #filterVal = this.#RULE('filterVal', () => {
    this.OR([{ ALT: () => this.SUBRULE(this.#filterIn) }, { ALT: () => this.SUBRULE(this.#atomicFilterVal) }]);
  });

  #filterIn = this.#RULE('filterIn', () => {
    this.CONSUME(t.LBracket);
    this.AT_LEAST_ONE_SEP({
      SEP: t.Comma,
      DEF: () => {
        this.SUBRULE(this.#atomicFilterVal);
      },
    });
    this.CONSUME2(t.RBracket);
  });

  #atomicFilterVal = this.#RULE('atomicFilterVal', () => {
    this.OR([
      {
        ALT: () => {
          this.CONSUME(t.LQuote);
          this.CONSUME(t.QuotedIdentifier);
          this.CONSUME2(t.RQuote);
        },
      },
      { ALT: () => this.CONSUME(t.Identifier) },
    ]);
  });

  #filterOp = this.#RULE('filterOp', () => {
    this.CONSUME(t.Colon);

    this.OPTION(() => {
      this.OR([
        { ALT: () => this.CONSUME(t.Minus, { LABEL: 'op' }) },
        { ALT: () => this.CONSUME(t.Plus, { LABEL: 'op' }) },
        { ALT: () => this.CONSUME(t.Equals, { LABEL: 'op' }) },
        { ALT: () => this.CONSUME(t.GreaterEq, { LABEL: 'op' }) },
        { ALT: () => this.CONSUME(t.LessEq, { LABEL: 'op' }) },
        { ALT: () => this.CONSUME(t.Greater, { LABEL: 'op' }) },
        { ALT: () => this.CONSUME(t.Less, { LABEL: 'op' }) },
      ]);
    });
  });
}

interface SearchParserOpts extends IParserConfig {}

export type SearchParser = SearchParserImpl;

export const createSearchParser = (opts?: SearchParserOpts) => {
  return new SearchParserImpl(lexerTokenDef, { recoveryEnabled: true, ...opts });
};
