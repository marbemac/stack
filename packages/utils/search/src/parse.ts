import type { ILexingResult, IRecognitionException, Lexer } from 'chevrotain';

import { SearchLexer } from './grammar/lexer.ts';
import { SearchParser } from './grammar/parser.ts';
import type { SearchString } from './types.ts';

let lexerSingleton: Lexer;
let parserSingleton: SearchParser;

export interface ParseQueryOpts<T extends InputType> {
  input: SearchString;
  inputType: T;
  lexer?: Lexer;
  parser?: SearchParser;
}

type InputType = 'searchQuery' | 'selectClause' | 'selectExpression' | 'fromClause' | 'whereClause' | 'whereExpression';

export const parseSearch = <T extends InputType>({
  input,
  inputType,
  ...rest
}: ParseQueryOpts<T>): {
  cst: ReturnType<SearchParser[T]>;
  errors: IRecognitionException[];
  lexResult: ILexingResult;
} => {
  let lexer = rest.lexer || lexerSingleton;
  if (!lexer) {
    lexer = lexerSingleton = new SearchLexer();
  }

  const lexResult = lexer.tokenize(input);

  let parser = rest.parser || parserSingleton;
  if (!parser) {
    parser = parserSingleton = new SearchParser();
  }

  // "input" is a setter which will reset the parser's state.
  parser.input = lexResult.tokens;

  const cst = parser[inputType]() as ReturnType<SearchParser[T]>;

  return { cst, errors: parser.errors, lexResult };
};
