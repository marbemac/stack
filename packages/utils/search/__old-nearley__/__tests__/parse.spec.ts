import nearley from 'nearley';
import { describe, expect, it } from 'vitest';

import grammar from '../grammar.ts';
import { isRawToken, parseSearchQuery } from '../parse.ts';
import type { SearchString } from '../types.ts';
import { caseGroups } from './cases.ts';

const prettyPrintParseResult = (input: string) => {
  const parser = new nearley.Parser(nearley.Grammar.fromCompiled(grammar));
  const res = parseSearchQuery(input as SearchString, { parser });

  const pretty: string[] = [];
  for (const r of res.tokens) {
    if (isRawToken(r)) {
      pretty.push(`${r.value}`);
      continue;
    }

    if (Array.isArray(r.val)) {
      if (r.negated) {
        pretty.push(`${r.key.value} NOT IN (${r.val.map(v => v.value).join(', ')})`);
      } else {
        pretty.push(`${r.key.value} IN (${r.val.map(v => v.value).join(', ')})`);
      }

      continue;
    }

    if (r.negated) {
      pretty.push(`!(${r.key.value} ${r.op} ${r.val.value})`);
    } else {
      pretty.push(`${r.key.value} ${r.op} ${r.val.value}`);
    }
  }

  return pretty;
};

describe.each(Object.entries(caseGroups))('%s', (_, cases) => {
  it.each(cases)('%s', (_, input) => {
    expect({
      input,
      result: prettyPrintParseResult(input),
    }).toMatchSnapshot();
  });
});
