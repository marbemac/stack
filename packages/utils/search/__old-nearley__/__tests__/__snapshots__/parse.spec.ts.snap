// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`freeform > freeform search mixed throughout 1`] = `
{
  "input": "freeform1 age:5 freeform2 search",
  "result": [
    "freeform1",
    "age = 5",
    "freeform2",
    "search",
  ],
}
`;

exports[`freeform > treats as a string 1`] = `
{
  "input": "freeform",
  "result": [
    "freeform",
  ],
}
`;

exports[`in filters > negation 1`] = `
{
  "input": "!user:[jane, john doe]",
  "result": [
    "user NOT IN (jane, john doe)",
  ],
}
`;

exports[`in filters > works 1`] = `
{
  "input": "user:[jane, john doe] release:[12.0]",
  "result": [
    "user IN (jane, john doe)",
    "release IN (12.0)",
  ],
}
`;

exports[`operators > < 1`] = `
{
  "input": "num_members:<5.0",
  "result": [
    "num_members < 5.0",
  ],
}
`;

exports[`operators > <= 1`] = `
{
  "input": "num_members:<=5.0",
  "result": [
    "num_members <= 5.0",
  ],
}
`;

exports[`operators > > 1`] = `
{
  "input": "num_members:>5.0",
  "result": [
    "num_members > 5.0",
  ],
}
`;

exports[`operators > >= 1`] = `
{
  "input": "num_members:>=5.0",
  "result": [
    "num_members >= 5.0",
  ],
}
`;

exports[`text filters > .-_ in keys 1`] = `
{
  "input": "payments.plan-type_is:free",
  "result": [
    "payments.plan-type_is = free",
  ],
}
`;

exports[`text filters > @.-_ in values 1`] = `
{
  "input": "email:jane@example.com type:public-company_x",
  "result": [
    "email = jane@example.com",
    "type = public-company_x",
  ],
}
`;

exports[`text filters > multiple 1`] = `
{
  "input": "plan:free age:5",
  "result": [
    "plan = free",
    "age = 5",
  ],
}
`;

exports[`text filters > negation 1`] = `
{
  "input": "!plan:free",
  "result": [
    "!(plan = free)",
  ],
}
`;

exports[`text filters > negation in quoted 1`] = `
{
  "input": "!user:"john doe"",
  "result": [
    "!(user = john doe)",
  ],
}
`;

exports[`text filters > quoted value 1`] = `
{
  "input": "user:"john doe"",
  "result": [
    "user = john doe",
  ],
}
`;

exports[`text filters > works 1`] = `
{
  "input": "plan:free",
  "result": [
    "plan = free",
  ],
}
`;
